File: chapter_2_linked_lists/js/2.2-returnKthToLast.js
const { SinglyLinkedList, Node } = require("./singlyLinkedList")

class SinglyLinkedList_2_2 extends SinglyLinkedList {
  /**
   * Returns the kth element from last node.
   * O(n) time complexity.
   * @param   {number}  index
   * @returns {Node}
   */
  kthToLast(index) {
    if (index > this.length || index < 0) throw new Error("index out of range")
    let curr = this.head
    let kth = this.head
    while (index) {
      index--
      curr = curr.next
    }
    while (curr) {
      curr = curr.next
      kth = kth.next
    }
    return kth
  }
}

const sll = new SinglyLinkedList_2_2()
const arr = [1, 2, 4, 8, 16, 32, 64]
for (let i of arr) sll.addNode(i)
sll.printList() // 64 -> 32 -> 16 -> 8 -> 4 -> 2 -> 1

let node = sll.kthToLast(1)
console.log(node.value) // 1
node = sll.kthToLast(3)
console.log(node.value) // 4
node = sll.kthToLast(5)
console.log(node.value) // 16

File: chapter_2_linked_lists/c/2.2-return_kth_to_last.c
#include "singly_linked_lists.h"

/**
 * Returns the kth element from last node.
 * O(n) time complexity.
 */
sll_node *kth_to_last(sll_node *head, int index)
{
    sll_node *curr, *kth;

    curr = kth = head;
    while (index)
    {
        curr = curr->next;
        index--;
    }
    while (curr)
    {
        curr = curr->next;
        kth = kth->next;
    }
    return kth;
}

int main()
{
    sll_node *head = NULL;
    int arr[] = {1, 2, 4, 8, 16, 32};
    int i, size = sizeof(arr) / sizeof(int);

    for (i = 0; i < size; i++)
        add_nodeint(&head, arr[i]);
    print_list(head); // 32 -> 16 -> 8 -> 4 -> 2 -> 1
    sll_node *kth = kth_to_last(head, 2);
    printf("%i\n", kth->n); // 2

    return 0;
}
